addons/sale/views/sale_order_line_views.xml

```diff
@@ -25,32 +25,70 @@
+                <field name="display_type" invisible="1"/>
                 <sheet>
                     <div class="oe_title">
                         <h1>
-                            <field name="display_name" readonly="1"/>
+                            <field name="display_name"/>
                         </h1>
                     </div>
                     <group>
                         <group>
+                            <field name="product_updatable" invisible="1"/>
                             <field name="order_id" readonly="1"/>
-                            <field name="product_id" readonly="1"/>
-                            <field name="name" readonly="1"/>
-                            <field name="product_uom_qty" readonly="1"/>
-                            <field name="qty_delivered" readonly="1"/>
-                            <field name="qty_invoiced"/>
-                            <field name="product_uom" readonly="1"/>
+                            <field name="product_uom" invisible="1"/>
+                            <field name="product_uom_category_id" invisible="1"/>
+                            <field name="product_uom_readonly" invisible="1"/>
+                            <field name="qty_delivered_method" invisible="1"/>
+                            <field name="state" invisible="1"/>
+                            <field name="product_id"
+                                attrs="{
+                                    'readonly': [
+                                        '&amp;',
+                                            ('id', '!=', False),
+                                            ('product_updatable', '=', False),
+                                    ],
+                                    'required': [('display_type', '=', False)],
+                                }"
+                                force_save="1"
+                                widget="many2one_barcode"
+                            />
+                            <field name="name"/>
+                            <field name="product_uom_qty"/>
+                            <field name="qty_delivered" 
+                                attrs="{
+                                    'invisible': [('state', 'not in', ['sale', 'done'])],
+                                    'readonly': [('qty_delivered_method', '!=', 'manual')],
+                                }"
+                            />
+                            <field name="qty_invoiced" attrs="{'invisible': [('state', 'not in', ['sale', 'done'])]}"/>
+                            <field name="product_uom"
+                                force_save="1"
+                                groups="uom.group_uom"
+                                attrs="{
+                                    'readonly': [('product_uom_readonly', '=', True)],
+                                    'required': [('display_type', '=', False)],
+                                }"
+                            />
```

Comment:
```suggestion
                            <field name="product_uom"
                                force_save="1"
                                groups="uom.group_uom"
                                attrs="{
                                    'readonly': [('product_uom_readonly', '=', True)],
                                    'required': [('display_type', '=', False)],
                                }"
                            />
                            <field name="product_uom" groups="!uom.group_uom" invisible="1"/>
```
Not 100% sure it's needed, but since `groups` now fully removes the element from the view (not only invisible) when the group is not enabled, it's good to still provide the field as invisible when the group is disabled (at least for feature groups like the uom one).